acvm-repo/brillig_vm/src/lib.rs

```diff
@@ -228,7 +258,209 @@ impl<'a, F: AcirField, B: BlackBoxFunctionSolver<F>> VM<'a, F, B> {
+            self.fuzzer_trace[index] += 1;
+        }
+    }
+    fn trace_binary_field_op(
+        &mut self,
+        op: &BinaryFieldOp,
+        lhs: MemoryValue<F>,
+        rhs: MemoryValue<F>,
+        result: MemoryValue<F>,
+    ) {
+        if self.fuzzing_active {
+            match op {
+                BinaryFieldOp::Add
+                | BinaryFieldOp::Sub
+                | BinaryFieldOp::Mul
+                | BinaryFieldOp::Div
+                | BinaryFieldOp::IntegerDiv => {}
+                BinaryFieldOp::Equals | BinaryFieldOp::LessThan | BinaryFieldOp::LessThanEquals => {
+                    let a = match lhs {
+                        MemoryValue::Field(a) => a,
+                        _ => {
+                            return;
+                        }
+                    };
+                    let b = match rhs {
+                        MemoryValue::Field(b) => b,
+                        _ => {
+                            return;
+                        }
+                    };
+                    let c = match result {
+                        MemoryValue::Field(..) => {
+                            return;
+                        }
+                        MemoryValue::U1(value) => value,
+                        _ => {
+                            return;
+                        }
+                    };
+                    let approach_index = self.branch_to_feature_map[&(
+                        self.program_counter,
+                        usize::MAX
+                            - 2
+                            - BigUint::from_bytes_be(&(b - a).to_be_bytes()).bits() as usize,
+                    )];
+                    let condition_index = self.branch_to_feature_map
+                        [&(self.program_counter, if c { usize::MAX } else { usize::MAX - 1 })];
+                    self.fuzzer_trace[condition_index] += 1;
+                    self.fuzzer_trace[approach_index] += 1;
+                }
+            }
+        }
+    }
 
+    fn trace_binary_int_op(
+        &mut self,
+        op: &BinaryIntOp,
+        lhs: MemoryValue<F>,
+        rhs: MemoryValue<F>,
+        bit_size: IntegerBitSize,
+        result: MemoryValue<F>,
+    ) {
+        if self.fuzzing_active {
+            match op {
+                BinaryIntOp::Add
+                | BinaryIntOp::Sub
+                | BinaryIntOp::Mul
+                | BinaryIntOp::Div
+                | BinaryIntOp::And
+                | BinaryIntOp::Or
+                | BinaryIntOp::Xor
+                | BinaryIntOp::Shl
+                | BinaryIntOp::Shr => {}
+                BinaryIntOp::Equals | BinaryIntOp::LessThan | BinaryIntOp::LessThanEquals => {
+                    let lhs_value: u128;
+                    let rhs_value: u128;
+                    match bit_size {
+                        IntegerBitSize::U1 => {
+                            lhs_value = match lhs.expect_u1() {
```

Comment:
Would it make sense to introduce a `MemoryValue::to_u128()` method? That said, this code seems to also be checking that the types of the arguments matches `bit_size` so using a method like `to_u128` would lose that, but I don't know if we need to do these assertions (maybe the binary op already checks this when it's built)